pkg/operator2/idp.go

```diff
@@ -292,6 +313,77 @@ func (c *authOperator) transportForCARef(ca configv1.ConfigMapNameReference, key
 	return transportFor("", caData, nil, nil)
 }
 
+func (c *authOperator) checkOIDCPasswordGrantFlow(
+	tokenURL, clientID string,
+	caRererence configv1.ConfigMapNameReference,
+	clientSecretReference configv1.SecretNameReference,
+) (bool, error) {
+	secret, err := c.secrets.Secrets("openshift-config").Get(clientSecretReference.Name, metav1.GetOptions{})
+	if err != nil {
+		return false, fmt.Errorf("couldn't get the referenced secret: %v", err)
+	}
+
+	// check whether we already attempted this not to send unneccessary login
+	// requests against the provider
+	if cachedResult, ok := oidcPasswordChecks[secret.ResourceVersion]; ok {
+		klog.V(4).Info("using cached result for OIDC password grant check")
+		return cachedResult, nil
+	}
+
+	clientSecret, ok := secret.Data["clientSecret"]
+	if !ok || len(clientSecret) == 0 {
+		return false, fmt.Errorf("the referenced secret does not contain the 'clientSecret' key")
+	}
+
+	transport, err := c.transportForCARef(caRererence, "ca")
+	if err != nil {
+		return false, fmt.Errorf("couldn't get a transport for the referenced CA: %v", err)
+	}
+
+	// prepare the grant-checking query, "openid" scope should be enough
+	query := url.Values{}
+	query.Add("client_id", clientID)
+	query.Add("client_secret", string(clientSecret))
+	query.Add("grant_type", "password")
+	query.Add("scope", "openid")
+	query.Add("username", "test")
+	query.Add("password", "test")
+	body := strings.NewReader(query.Encode())
+
+	// prepare the request, require application/json for simple parsing the response
+	req, err := http.NewRequest("POST", tokenURL, body)
+	if err != nil {
+		return false, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	reqBytez, err := httputil.DumpRequest(req, true)
+	klog.V(1).Infof("teh bytes be: %s", reqBytez)
+
+	client := &http.Client{Transport: transport}
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	respJSON := json.NewDecoder(resp.Body)
+	respMap := map[string]interface{}{}
+	if err = respJSON.Decode(&respMap); err != nil {
+		return false, fmt.Errorf("failed to decode response from the OIDC server: %v", err)
+	}
+
+	if errVal, ok := respMap["error"]; ok {
+		oidcPasswordChecks[secret.ResourceVersion] = errVal == "invalid_grant" // wrong password, but password grants allowed
+	} else {
+		_, ok = respMap["access_token"] // in case we managed to hit the correct user
```

Comment:
Hmm, in case l/p test/test works?

I can see how this would be a good manual test for whether a password check is supported. Is this best practice for detecting oidc capability though? My naive expectation would be that it wouldn't be necessary to authenticate to discover what types of authentication were supported, but I'll openly admit that I don't know auth conventions.

Reply:
> Hmm, in case l/p test/test works

Yes

> I can see how this would be a good manual test for whether a password check is supported. Is this best practice for detecting oidc capability though?

I originally thought that we could simply depend on the discovery endpoint, specifically on the `grant_types_supported` field, but this only seems to be set by Google, neither GitLab nor Azure expose the field, GitHub does not have the endpoint at all. And even if this field was set in discovery, the password grant flow can still be disabled on some oauth clients.

I think that leaves us in the uncomfortable position for having to check for the flow ourselves :disappointed: